### PR TITLE
Add SSL Client support

### LINKED ISSUE #61: Talk SSL to the auth server
My auth server *only* talks HTTPS. Right now, I use stunnel in client mode to tunnel the connection. This pulls in quite a lot of dependencies.

My feature request/RFC: hook up a small SSL library like PolarSSL or the smaller cyassl. This could be  enabled on compile time. Hooking up [cyassl is quite straightforward](http://www.yassl.com/yaSSL/Docs-cyassl-manual-3-getting-started.html).

Would there be interest in such a feature? Any ideas on that?

### PR DESCRIPTION
**Pull request by mhaas**
This pull request adds optional SSL client support when talking to the auth server. This is useful if your auth server only speaks HTTPS or if you just want to use HTTPS in a post-snowden world.

The patch is tested lightly: it works with my server. The code is heavily based on the [http://www.yassl.com/yaSSL/Docs-cyassl-manual-11-ssl-tutorial.html](CyaSSL tutorial). No guarantee regards "security" given :)

Fixes #61

### CODE REVIEW

**Comment by mhaas** [comment]
Two problems:

* ca-certificates are not loaded, so SSLNoPeerVerification is required
* in conf.c, the debug logs fire although they are #ifndef'd with USE_CYASSL

File: src/conf.c [hunk 1]
```diff
@@ -96,6 +97,8 @@ typedef enum {
 	oTrustedMACList,
 	oHtmlMessageFile,
 	oProxyPort,
+	oSSLNoPeerVerification,
```
Comment (acv) [inline src/conf.c]: I would suggest a key to disable the use of HTTPS globally when talking to AuthServer for testing purpose.

File: src/conf.h [hunk 2]
```diff
@@ -67,6 +67,8 @@
 #define DEFAULT_AUTHSERVMSGPATHFRAGMENT "gw_message.php?"
 #define DEFAULT_AUTHSERVPINGPATHFRAGMENT "ping/?"
 #define DEFAULT_AUTHSERVAUTHPATHFRAGMENT "auth/?"
+#define DEFAULT_AUTHSERVSSLCERTPATH "/etc/ssl/certs/"
+#define DEFAULT_AUTHSERVSSLNOPEERVER 0
```
Comment (acv) [inline src/conf.h]: Would suggest making sure the name clearly indicate whether validation will be done or not, as it is, I don't know what 0 means here? Is default on or default off?

File: src/simple_http.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 59 Temple Place - Suite 330        Fax:    +1-617-542-2652       *
+ * Boston, MA  02111-1307,  USA       gnu@gnu.org                   *
+ *                                                                  *
+ \********************************************************************/
+
+int http_get(const int, char*);
+
+#ifdef USE_CYASSL
```
Comment (acv) [inline src/simple_http.h]: `HAVE_CYASSL` might be a better name?

File: src/simple_http.c [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+			return -1;
+		}
+	} while (!done);
+
+	close(sockfd);
+
+	buf[totalbytes] = '\0';
+	debug(LOG_DEBUG, "HTTP Response from Server: [%s]", buf);
+
+    CyaSSL_free(ssl);
```
Comment (acv) [inline src/simple_http.c]: White space.

File: src/simple_http.c [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+	if (sockfd == -1) {
+		/* Could not connect to server */
+		debug(LOG_ERR, "Could not open socket to server!");
+		return -1;
+	}
+
+
+	/* Create CYASSL object */
+	CYASSL* ssl;
```
Comment (acv) [inline src/simple_http.c]: White space is weird... But more importantly there should be very clear logging of the certificate validation failure along with some info about options (fix cert path, disable check) or at least a pointer to config.

File: src/ping_thread.c [hunk 6]
```diff
@@ -168,61 +167,20 @@ ping(void)
...
-
-	debug(LOG_DEBUG, "HTTP Response from Server: [%s]", request);
+	int res = 0;
+	#ifdef USE_CYASSL
+	if (auth_server->authserv_use_ssl) {
+		res = https_get(sockfd, request);
+	} else {
+		res = http_get(sockfd, request);
+	}
+	#endif
```
Comment (acv) [inline src/ping_thread.c]: Can use `#else`

File: src/centralserver.c [hunk 7]
```diff
@@ -330,9 +288,22 @@ int _connect_auth_server(int level) {
...
+		int port = 0;
+		#ifdef USE_CYASSL
+		if (auth_server->authserv_use_ssl) {
+			debug(LOG_DEBUG, "Level %d: Connecting to SSL auth server %s:%d", level, hostname, auth_server->authserv_ssl_port);
+			port = htons(auth_server->authserv_ssl_port);
+		} else {
+			debug(LOG_DEBUG, "Level %d: Connecting to auth server %s:%d", level, hostname, auth_server->authserv_http_port);
+			port = htons(auth_server->authserv_http_port);
+		}
+		#endif
```
Comment (acv) [inline src/centralserver.c]: Again, `#else` can be used.

**Comment by acv** [comment]
Stylistic thing but generally, it's best to do white space / formatting fix separately to minimize code to review ;-)

File: src/simple_http.h (hunk 3 above)
Comment (mhaas) [inline src/simple_http.h]: HAVE_CYASSL is set by AC_HAVE_LIBRARY. In the current patch, autoconf only uses AC_HAVE_LIBRARY(cyassl) if --enable-cyassl is passed, so we can indeed use that. However, I wanted to have a distinction between "cyassl is available" and "we actually want to use ssl".

**Comment by mhaas** [comment]
@avc Sorry about the white space. I tried to minimize his - or are you talking about keeping white space and actual code PRs separate?


Thanks for the thorough review!

**Comment by acv** [comment]
Yeah, the latter. White space PRs as separate.

File: src/simple_http.h (hunk 3 above)
Comment (acv) [inline src/simple_http.h]: That makes sense.

**Comment by mhaas** [comment]
Thanks. There are some other proiblems:

* I can't load SSL certs, the call fails with E_COMPRESS for some reason. Maybe I'm missing zlib?!
* I might have to enable domain name verification separately

**Comment by mhaas** [comment]
Do you need me to re-do the pull request re whitespace or can we let it slip one time? ;)

**Comment by acv** [comment]
Don't worry about it.

File: src/conf.h (hunk 2 above)
Comment (mhaas) [inline src/conf.h]: The option is off (0) by default. The name SSLNoPeerVerification - you have to actively turn it on as a reminder that this is potentially bad.
Is 4545c6a better?

File: src/ping_thread.c (hunk 6 above)
Comment (mhaas) [inline src/ping_thread.c]: How so? Even when compiled with USE_CYASSL, we still have to check if the authserver is configured to use ssl.

File: src/conf.c (hunk 1 above)
Comment (mhaas) [inline src/conf.c]: You can simple set SSLAvailable to No in the auth server section. Clients will be redirected to HTTP then, but if it's just for testing, that should be ok.

File: src/simple_http.c (hunk 5 above)
Comment (mhaas) [inline src/simple_http.c]: Better in 59ef4b3?

File: src/conf.c (hunk 1 above)
Comment (mhaas) [inline src/conf.c]: Added a note in wifidog.conf: 4f634ed

File: src/simple_http.c (hunk 5 above)
Comment (acv) [inline src/simple_http.c]: Yes.

**Comment by databeille** [comment]
Much wanted feature ! :+1:

**Comment by mhaas** [comment]
CyaSSL/my code won't load ca-certificates (OpenWrt BB) properly. The set_verify_locations call returns E_COMPRESS.

Some hypotheses:

    I somehow managed to compile without zlib
    There is a problem with the hash symlinks: OpenSSL c_rehash changed from MD5 to SHA1. However, CyaSSL apparently does not care about the naming conventions
    CyaSSL gets confused by the symlinks

Possible workarounds if I can't get it fixed:

    Let user specify single root CA
    Concatenate all CAs into single file

@databeille If you want, you can help test.

**Comment by acv** [comment]
I suspect the issue would be with the cert files in the directory you're reading from.

**Comment by databeille** [comment]
@mhaas : having a look to your code, not being a C++ dev :/

As I use openSSL in BB, I have to run a script at startup to make it work flawlessly with CAcerts. May it help ?

<pre>
#!/bin/sh
#original source : https://forum.openwrt.org/viewtopic.php?id=50661
OPENSSL="/usr/bin/openssl"
CERTDIR="/etc/ssl/certs"
TMPCERTDIR="/tmp$CERTDIR"
FILE="$1"

go () {
makedir $TMPCERTDIR
# Install openssl-util if need
[ ! -f ${OPENSSL} ] && opkg update && opkg install openssl-util
if [ "$FILE" = "" ]
then
        rm `find /tmp$CERTDIR -maxdepth 1 -type l -print` > /dev/null 2>&1
        for CERTFILE in ${CERTDIR}/*
                do
                        makehash
                done
else
        CERTFILE=$FILE
        CERTDIR=""
        makehash
fi
}
makedir () {
build=""
[ "${1:0:1}" = "/" ] && build="/"
        for dir in $(echo "$1" | grep -Eoi "[^/]*")
                do
                        build="$build$dir/"
                        mkdir -p "$build"
                done
}

makehash () {
        # create symbolic link from hash
#        echo -en "Certificate ${CERTFILE##*/}\n  generating hash: "
        HASH=$(${OPENSSL} x509 -hash -noout -in "${CERTFILE}")
#        echo "$HASH"

        # handle hash collision
        SUFFIX=0
        while [ -h "${TMPCERTDIR}/${HASH}.${SUFFIX}" ]; do
                let "SUFFIX += 1"
        done

#        echo "  linking ${HASH}.${SUFFIX} -> ${CERTFILE##*/}"
        ln -s "$CERTDIR/${CERTFILE##*/}" "${TMPCERTDIR}/${CERTFILE##*/}"
        ln -s "$CERTDIR/${CERTFILE##*/}" "${TMPCERTDIR}/${HASH}.${SUFFIX}"
}

</pre>

Basically, the script generates "hashes" in /tmp/etc/ssl/certs for certs in /etc/ssl/certs.
When needed, I tell the program where to find them : 
<pre>
wget -qO- $1 --secure-protocol=PFS --ca-directory=$SSL_CERT_DIR
</pre>

**Comment by acv** [comment]
The configure.in should be extended to check for `-lcyassl`. If you install wolfSSL (the new version) the headers are found but you need `-lwolfssl` to link.

File: src/centralserver.c [hunk 8]
```diff
@@ -326,9 +284,22 @@ int _connect_auth_server(int level) {
...
+		#ifndef USE_CYASSL
 		debug(LOG_DEBUG, "Level %d: Connecting to auth server %s:%d", level, hostname, auth_server->authserv_http_port);
+		port = htons(auth_server->authserv_http_port);
+		#endif
```
Comment (acv) [inline src/centralserver.c]: You're running `htons()` twice. On little-endian architectures, this corrupts the port. On big-endian it does nothing.

Broke for me on amd64 / Ubuntu.

**Comment by acv** [comment]
While trying to test this, it seems that a lot of the certs that ship with Ubuntu will cause cyassl to die with error -148, something about an unknown OID sum. I tried to winnow them out but after 4, I stopped ;-)

**Comment by acv** [comment]
Turns out I needed --enable-ecc in cyassl to use the system certs.

**Comment by acv** [comment]
I now get this error on every https_get:

    CyaSSL_send failed: record layer length error

**Comment by acv** [comment]
Turns out that if I properly configure my Ubuntu test box with a work HTTPS server, I don't get that record layer length error... The more you know!

**Comment by acv** [comment]
When done reading an HTTPS response, I get this log entry:

    Finished reading from server. CyaSSL message: error state on socket

I expect it's spurious.

**Comment by acv** [comment]
Rather then messing about with a pull request, I emailed you a patch for the issues I found so far.

**Comment by mhaas** [comment]
@acv 

* wolfssl: 76a6fb5
* error -184: 5ef25c0
* double negation: 86c8247
* regarding "Finished reading from server. CyaSSL message: error state on socket", I am not sure how to handle this. I don't see a way to differentiate between a clean shutdown and a "connection closed by peer", although there probably is one

**Comment by acv** [comment]
Seems OK... Except that instead of COMPRESS_E, I got ASN_UNKNOWN_OID_E when I needed `--enable-ssl`

I got no errors with these options:

    --enable-ssl --with-libz --enable-aesgcm --enable-aesni

Everything else stock.

**Comment by acv** [comment]
For the "error state on socket" error, I'd suppress its logging. If we can't do any better, we might as well avoid ourselves support questions.

**Comment by mhaas** [comment]
@acv, thanks! I have to verify COMPRESS_E then.

Testing on my VM right now, but apparently wifidog linked against host libc. Time to fix my build.

**Comment by mhaas** [comment]
You were right, it is ASN_UNKNOWN_OID_E. I mixed up -184 and -148.

**Comment by acv** [comment]
Sweet, so we have a root cause (ECC root certs) and a fix (`--enable-ecc`).

**Comment by mhaas** [comment]
Removed the "error state on socket" message.

cyassl with --enable-ecc works for me now.

I think we can merge. I didn't test against known broken servers (e.g. expired, wrong hostname), though.

**Comment by acv** [comment]
I would suppress the logging when `numbytes == 0` in the read loop personally but I think it's merge ready now otherwise.

**Comment by acv** [comment]
I checked the hostname mismatch case and it prints a meaning full error about peer mismatch.

**Comment by mhaas** [comment]
Good. You pull the trigger, we share the blame.

**Comment by mhaas** [comment]
FYI, i requested that OpenWrt enable ECC: https://dev.openwrt.org/ticket/19188#ticket

**Comment by acv** [comment]
Sweet.

**Comment by databeille** [comment]
Do you have : 
(simple_http.c:256) CyaSSL_send failed: don't have enough data to complete task
when trying to ping authserver ?

build for openwrt BB with cyassl 3.3.0 or cyassl 3.3.2 master

**Comment by mhaas** [comment]
No. Can you post your config? What platform?

**Comment by acv** [comment]
That sounds like connect to a non SSL server with SSL.

This happens when the server response is not the expected size on the TLS handshake:

```C
    if (helloSz < RECORD_HEADER_SZ + HANDSHAKE_HEADER_SZ + CLIENT_HELLO_FIRST)
	        return INCOMPLETE_DATA;
```

**Comment by databeille** [comment]
@mhaas I have a ar71xx router, running openwrt BB.

I use https://github.com/indutny/bud as TLS terminator, will try with a basic apache setting to check.

**Comment by mhaas** [comment]
On 03/30/2015 02:33 PM, databeille wrote:
> @mhaas <https://github.com/mhaas> I have a ar71xx router, running
> openwrt BB.
> 
> I use https://github.com/indutny/bud as TLS terminator, will try with a
> basic apache setting to check.

Does it work with a regular browser? Perhaps the problem lies between
the terminator and your webserver.

**Comment by databeille** [comment]
Yes, usually, everything works perfectly :)
With Apache handling SSL-TLS, wifidog-tls works, I got a "Pong".
bud-tls returns many different error codes with wifidog-tls, the "weirdest" being : "no shared cipher".

**Comment by mhaas** [comment]
Hm, ok. It would be good if you made a ticket with as much information (logs) as you can and I will take a look :)

Am 30. März 2015 16:20:51 MESZ, schrieb databeille <notifications@github.com>:

**Comment by acv** [comment]
Can you run Qualys' SSL tester against your bud-tls endpoint? https://www.ssllabs.com/ssltest/

**Comment by databeille** [comment]
@acv : yes, is works great !
I try to configure it less severe (+SSL2 +SSL3 +UseClientCypherSuite +...), but nothing good for now...

**Comment by acv** [comment]
Wifidog is hard coded to to TLS 1.x or greater only so SSL2/SSL3 is not supported.

**Comment by databeille** [comment]
wifidog-tls works great with Apache+mod_ssl, Apache+stud (https://github.com/bumptech/stud/ ; my previous terminator, not maintained anymore) but miserably fail with bud-tls.
I give up for now.

**Comment by mhaas** [comment]
@databeille I assume you have the cyassl source handy? In examples/client, there is a test program. Execute something like this:

    ./client -h yourserver.xzy -p 443

So we can see if it's a wifidog problem or a cyassl+bud-tls problem.

**Comment by databeille** [comment]
It is not related to wifidog.
cyassl's ./client script fails too (master branch).

**Comment by databeille** [comment]
My last post about this problem : "INCOMPLETE_DATA" being wide (cases in cyassl/src/tls.c and cyassl/src/internal.c), the error code returned is related to : 
<pre>    /* make sure can read the message */
    if (*inOutIdx + size > totalSz)
        return INCOMPLETE_DATA;</pre>
in src/internal.c.

Will know push an issue to bud-tls.
